runtime/lib/ttnn/program.cpp

```diff
@@ -377,33 +398,69 @@ run(::tt::target::ttnn::Operation const *op, ::ttnn::Device &device,
   }
 }
 
-void runProgram(::ttnn::Device &device,
-                ::tt::target::ttnn::Program const *program,
-                std::vector<::ttnn::Tensor *> const &inputs,
-                std::vector<::ttnn::Tensor *> const &outputs) {
+std::vector<Tensor> runProgram(::ttnn::Device &device,
+                               ::tt::target::ttnn::Program const *program,
+                               std::vector<::ttnn::Tensor *> const &inputs) {
   std::unordered_map<std::uint32_t, ::ttnn::Tensor *> liveTensors;
   std::list<::ttnn::Tensor> tensorPool;
 
   int inputIndex = 0;
   assert(program->inputs()->size() == inputs.size() &&
          "Mismatch between program inputs and input tensors");
+
+  // convert inputs to the desired layout/memory config
   for (::tt::target::TensorRef const *input : *program->inputs()) {
+    const ::ttnn::Tensor *inputTensor = inputs[inputIndex++];
+    const ::tt::target::MemoryDesc *inputDesc =
+        input->desc()->layout()->memory_desc();
+    ::ttnn::Tensor layoutUpdatedInputTensor =
+        updateTensorMemoryConfig(device, *inputTensor, inputDesc);
+    tensorPool.emplace_back(std::move(layoutUpdatedInputTensor));
     auto [iter, inserted] =
-        liveTensors.try_emplace(input->global_id(), inputs[inputIndex++]);
+        liveTensors.try_emplace(input->global_id(), &tensorPool.back());
     assert(inserted && "Duplicate input tensor");
   }
 
-  int outputIndex = 0;
-  assert(program->outputs()->size() == outputs.size() &&
-         "Mismatch between program outputs and output tensors");
-  for (::tt::target::TensorRef const *output : *program->outputs()) {
-    auto [iter, inserted] =
-        liveTensors.try_emplace(output->global_id(), outputs[outputIndex++]);
-    assert(inserted && "Duplicate output tensor");
-  }
-
   for (::tt::target::ttnn::Operation const *op : *program->operations()) {
     run(op, device, liveTensors, tensorPool);
   }
+
+  // convert outputs to the desired layout/memory config
+  // then convert them to type erased tensors and return
+  std::vector<Tensor> outputs;
+  for (::tt::target::TensorRef const *output : *program->outputs()) {
+    size_t output_id = output->global_id();
+    TT_FATAL(liveTensors.contains(output_id),
+             "Program output tensor with global_id {} not found in liveTensors",
+             output_id);
+    const ::ttnn::Tensor *outputTensor = liveTensors.at(output_id);
+    const ::tt::target::MemoryDesc *outputDesc =
+        output->desc()->layout()->memory_desc();
+    ::ttnn::Tensor layoutUpdatedOutputTensor =
+        updateTensorMemoryConfig(device, *outputTensor, outputDesc);
+    outputs.push_back(toTypeErasedTensor(std::move(layoutUpdatedOutputTensor)));
```

Comment:
this way, output tensors will be either on the host or on device, depending on the `outputDesc` in the flatbuffer?

how will the FE runtime move it to host?

Reply:
That's correct it could be on host or device depending on outputDesc. If FE wants to move to host, with the current implementation it would need to have a program with output desc host. 

If desired I can overload toLayout that accepts some layout descriptors to decouple it from the flatbuffer.